src/librustc_codegen_ssa/back/link.rs

```diff
@@ -460,6 +460,101 @@ fn link_staticlib<'a, B: ArchiveBuilder<'a>>(sess: &'a Session,
     }
 }
 
+
+/// rust-lang/rust#61539: bugs in older versions of GNU `ld` cause problems that
+/// are readily exposed under our default setting of disabling PLT (PR
+/// rust-lang/rust#54592). Heuristically detect and warn about this.
+fn check_for_buggy_ld_version(sess: &Session,
+                              program_name: &Path,
+                              flavor: LinkerFlavor,
+                              crate_type: config::CrateType) {
+    debug!("check_for_buggy_ld_version: \
+            program_name: {:?} flavor: {:?} crate_type: {:?}",
+           program_name, flavor, crate_type);
+
+    match crate_type {
+        config::CrateType::Dylib |
+        config::CrateType::ProcMacro => (),
+
+        // FIXME: should we include CrateType::Cdylib in the warning? It is not
+        // clear why we haven't seen it there.
+        config::CrateType::Cdylib => return,
+
+        // Static objects won't run into this (unless they load a dynamic
+        // object, which this heuristic is not attempting to detect).
+        config::CrateType::Executable |
+        config::CrateType::Rlib |
+        config::CrateType::Staticlib => return,
+    };
+
+    let mut version_cmd = Command::new(program_name);
+    match flavor {
+        LinkerFlavor::Gcc => {
+            // run `gcc -v -Xlinker --version` to query gcc for version info of underlying linker
+            version_cmd.args(&["-v", "-Xlinker", "--version"]);
+        }
+        LinkerFlavor::Ld => {
+            // run `ld --version`
+            version_cmd.args(&["--version"]);
+        }
+        LinkerFlavor::Msvc |
+        LinkerFlavor::Em |
+        LinkerFlavor::Lld(..) |
+        LinkerFlavor::PtxLinker => {
+            // Not GNU ld, so don't bother inspecting version.
+            return;
+        }
+    }
+    let version_check_invocation = format!("{:?}", &version_cmd);
+    debug!("check_for_buggy_ld_version version_check_invocation: {:?}",
+           version_check_invocation);
+    let output = match version_cmd.output() {
+        Err(_) => {
+            sess.warn(&format!("Linker version inspection failed to execute: `{}`",
+                               version_check_invocation));
+            return;
+        }
+        Ok(output) => output,
+    };
+    let out = String::from_utf8_lossy(&output.stdout);
+
+    debug!("check_for_buggy_ld_version out: {:?}", out);
+
+    let parse = |first_line: &str| -> Option<(u32, u32)> {
+        let suffix = first_line.splitn(2, "GNU ld version ").last()?;
```

Comment:
on all of the machines I have ld returns `GNU ld (GNU Binutils .*) X.Y.Z` so this would give you e.g.

```
("GNU ld (GNU Binutils for Ubuntu) 2", "26")
```

for e.g. ubuntu’s ld, if I’m reading this code right.

FWIW `-v` and `-V` return different values as well:

```
$ ld -v
GNU ld (GNU Binutils for Ubuntu) 2.26.1
$ ld -V
GNU ld (GNU Binutils for Ubuntu) 2.26.1
  Supported emulations:
   elf_x86_64
   elf32_x86_64
   elf_i386
   elf_iamcu
   i386linux
   elf_l1om
   elf_k1om
   i386pep
   i386pe
$ ld --version
GNU ld (GNU Binutils for Ubuntu) 2.26.1
Copyright (C) 2015 Free Software Foundation, Inc.
This program is free software; you may redistribute it under the terms of
the GNU General Public License version 3 or (at your option) a later version.
This program has absolutely no warranty.
$ ld -v # on recent nixos channel
GNU ld (GNU Binutils) 2.31.1
```

Reply:
Hmm, yeah I was worried about this sort of thing.

I could instead just search for a three character sequence matching the regexp ` [0-9]\.` and treat that as marking the beginning of the version string. (Yes that could get tripped up by pathological cases but I think its a plausible route forward.)

Reply:
Okay I decide to search for a version string that contains both `GNU ld` and `2.x.y-zzz` somewhere in it, and then the `2.x.y-zzz` is parsed into just `(2, x)`.